lib/ex_doc/autolink.ex

```diff
@@ -509,25 +510,17 @@ defmodule ExDoc.Autolink do
     end
   end
 
-  defp maybe_warn({kind, module, name, arity}, config) do
+  defp maybe_warn(ref, config, visibility, metadata) do
     skipped = config.skip_undefined_reference_warnings_on
     file = Path.relative_to(config.file, File.cwd!())
     line = config.line
 
     unless Enum.any?([config.id, config.module_id, file], &(&1 in skipped)) do
-      warn({kind, module, name, arity}, file, line, config.id)
+      warn(ref, {file, line}, config.id, visibility, metadata)
     end
   end
 
-  defp warn({kind, module, name, arity}, file, line, id) do
-    message =
-      "documentation references #{kind} #{inspect(module)}.#{name}/#{arity}" <>
-        " but it is undefined or private"
-
-    warn(message, file, line, id)
-  end
-
-  defp warn(message, file, line, id) do
+  defp warn(_ref, {file, line}, id, _visibility, %{message: message}) do
```

Comment:
I think this could be a warn/3:

```suggestion
  defp warn(message, {file, line}, id) do
```

and then all the other warn/5 clauses call it.

Reply:
done